Pyrun template

sdomme opened the pull request:
We can now use [PyRun](http://www.egenix.com/products/python/PyRun/) as Python distribution. Miniconda remains default.

sdomme committed:
With this changes in spec TEMPLATE it is possible to build a binary RPM completely in userspace without the need of "opt-permission-change-rpm-voodoo"

sdomme committed:
Describe a possible usage of snakepit and rpmbuild

sdomme committed:
gitignore

sdomme committed:
pyrun extension template with optional libraries linkage

sdomme committed:
you can build a pyrun spec file

sdomme committed:
Implement a feature to build spec files with pyrun distribution

sdomme committed:
fix cram tests

sdomme committed:
include templates and extensions to the build

sdomme committed:
you can choose to use pyrun as distribution. miniconda is default

sdomme committed:
some refactoring on template generation and build number construction

sdomme committed:
fix cram

sdomme committed:
Explain PyRun integration in README

sdomme committed:
include *.spec

sdomme committed:
spec files for miniconda and PyRun

schlomo commented on src/main/python/snakepit/main.py:
```diff
@@ -84,6 +84,43 @@ def custom_output_filename(filename, output_directory):
+                                               yaml_spec['conda_dist_version'],)
+    elif distribution == "pyrun":
+        build_number = "{0}_pyrun_{1}".format(yaml_spec['build'],
+                                              yaml_spec['pyrun_dist_version'])
```
The actualy Python version (2.7 or 3.4) is missing from the PyRun build path.

I find it important that we also make the Python version transparent on at least some level.

sdomme replied:
RPM example:

gaius-128.0-0_pyrun_2.1.1.x86_64.rpm

Don't want to bloat this with a one more version.

@schlomo Some idea where else?

schlomo replied:
I would prefer to add `_py2.7` to the Release. I would put it in the end because it is least relevant but building with a higher Python version will still give you an upgrade path.

sdomme replied:
done

schlomo commented on README.md:
````diff
@@ -62,6 +63,47 @@ To install, do:
+
+Example:
+```
+libraries: {'/usr/lib64/libssl.so':'libssl.so.1.0.0', '/usr/lib64/libcrypto.so':'libcrypto.so.1.0.0'}
````
Why not simply download the PyRun Source and compile it against the build OS as part of the rpmbuild? That way you don't need this library linking at all and you can make sure that the PyRun environment actually fits the target OS.

sdomme replied:
Good idea, but there are some software requirements for PyRun compilation in the documentation, so we need to take care of it too.

schlomo replied:
RPM BuildRequires should solve that problem.

sdomme replied:
Ok. Its a bigger structural change in the template, so I would like to test it on a separate branch next time.

schlomo replied:
The cost of pushing this change into the next round is that then you have to deal with obsoleting the library stuff from already installed systems.

Another risk: By including links from `/*/lib64/*` you actually pin the package to be 64bit without hard-coding that in the spec file.

Either be platform-independant or hard-code your target platform everywhere.

sdomme replied:
changed to the "compile-PyRun-source"-Template

schlomo commented on src/main/python/snakepit/templates/TEMPLATE-PyRun.spec:
```diff
@@ -0,0 +1,55 @@
+# use pip to install {{ pypi_package_name }}
+{{ pypi_package_name }}/bin/pip install {{ extra_pip_args }} --no-compile --global-option build_scripts --global-option "--executable=/opt/{{ pypi_package_name }}/bin/python" {{ pypi_package_name }}=={{ pypi_package_version }}
+#cleanup
+{{ pypi_package_name }}/bin/python -m pip uninstall pip setuptools -y
```
Why is the way to run pip different between this uninstall line and the previous install line?

sdomme replied:
You're right. "pip uninstall pip setuptools -y" would do the job.

sdomme replied:
done

schlomo commented on src/main/python/snakepit/templates/TEMPLATE-PyRun.spec:
```diff
@@ -0,0 +1,55 @@
+%install
+# create /opt/{{ pypi_package_name }} in buildroot
+install -m 755 -d %{buildroot}/opt/{{ pypi_package_name }}
+# copy the built {{ conda_dist_flavour }} env into the buildroot
```
conda_dist_flavour???

sdomme replied:
will fix this bug

sdomme replied:
done

schlomo commented on src/main/python/snakepit/templates/TEMPLATE-PyRun.spec:
```diff
@@ -0,0 +1,55 @@
+# clean, just in case
+rm -rf {{ pypi_package_name }}
+{% block libraries %}{% endblock %}
+curl -O https://downloads.egenix.com/python/install-pyrun
```
Alternative Option: Download source and compile.

sdomme replied:
done

schlomo commented on src/main/python/snakepit/templates/TEMPLATE-PyRun.spec:
```diff
@@ -0,0 +1,55 @@
+curl -O https://downloads.egenix.com/python/install-pyrun
+bash install-pyrun {{ pypi_package_name }}
+# use pip to install {{ pypi_package_name }}
+{{ pypi_package_name }}/bin/pip install {{ extra_pip_args }} --no-compile --global-option build_scripts --global-option "--executable=/opt/{{ pypi_package_name }}/bin/python" {{ pypi_package_name }}=={{ pypi_package_version }}
```
All that could be also part of the %install phase and go directly into the %buildroot area. Save one copy :-)

sdomme replied:
Not if we download source and compile :-)
This stuff has a touch of building something, therefore %build directive :-)

schlomo replied:
What purpose does this separation between `%build` and `%install` serve? IMHO not much and we should be lenient about optimizing between those two stages.

sdomme replied:
all in %install now

schlomo commented:
Very nice, PyRun looks like an even better solution than miniconda.

schlomo commented:
BTW, I found on this info: http://www.egenix.com/products/python/PyRun/#Installation

**Complete Python runtime installation with dependencies**
If you want to install PyRun together with pip and all your project dependencies, use the -r requirements.txt option:

```
./install-pyrun -r requirements.txt targetdir
```

Maybe this is an even easier way to install PyRun and the target module via pip?

sdomme commented:
```
./install-pyrun -r requirements.txt targetdir
```
Yes it's easier, but with this method you can't modify shebang in the scripts like we do it with
```
bin/pip install --no-compile --global-option build_scripts --global-option "--executable=/opt/{{ pypi_package_name }}/bin/python" {{ pypi_package_name }}
```

Setuptools would inject the interpreter inside the build directory by default 

It would break the connection to the python interpreter inside /opt/... .

sdomme committed:
PyRun in the comment for PyRun template

sdomme committed:
uninstall pip and setuptools the pip way.

sdomme commented:
In my opinion the big advantage of the PyRun method is, after building the spec file you can simply run "buildrpm -bb software.spec" no matter where. No svn2rpm, no make-opt-writable

schlomo commented:
I am not sure that the `rpmbuild -bb *.spec` method works everywhere because PyRun itself is platform dependant (32/64 bit, Arm vs. Intel ...).

Actually downloading the PyRun source and building that will be much more portable and only rely on the fact that `rpmbuild -bb` is run on the target platform. This is actually the basic assumption that all packages take since we don't use cross-building or cross-compiling.

sdomme committed:
add python version to the Release

sdomme committed:
must contain minor version

sdomme committed:
This PyRun template:
- download, compile and install PyRun to buildroot
- download and install setuptools and pip
- install the software

sdomme committed:
no more complexity with libraries linage with the new PyRun template

sdomme committed:
Build setuptools and pip in BUILD-dir, so we can skip the cleanup task for it

sdomme committed:
Readme

sdomme committed:
PyRun Python version can be specified with  pyrun_pythonfullversion argument in snakepit-yaml. Default is 2.7.10

sdomme committed:
Use pyrun_dist_version for PyRun download. Default is 2.1.1

schlomo commented:
Very nice! Looks IMHO much cleaner now.

One question remains: Why do we have to use 2 different configuration variables for the Python version (e.g. 2.7 or 3.4) for miniconda and PyRun? Why not simply have one variable for that information (e.g. `python_version`) that will be used by both the miniconda and the PyRun code appropriately. The goal would be to make switching between miniconda and PyRun as simple as possible, just change the "driver" setting without changing anything else.

sdomme committed:
If build PyRun with Python 3, PyRun creates only python3 symlinks to the interpreter.

sdomme commented:
PyRun needs 3-postion-python-version, otherwise the compilation breaks. Also I didn't see where miniconda needs python version. I think it's not implemented to choose it by snakepit. So I think it didn't hurt to rename it.

schlomo commented:
According to https://repo.continuum.io/miniconda/ also miniconda uses the 3-position-python-version.

sdomme commented:
It's miniconda version not python.

schlomo commented:
@sdomme good point - I did not look closely enough.

Then maybe the other way is actually better: The user should only be able to specify the Python major version (2 or 3) and then snakepit takes the "latest" available for the chosen "driver"? In any that is what happens with Miniconda, so why not deliver the same behavior with PyRun?

sdomme commented:
In my opinion it is a cool feature, if you can specify a minor version (even if it's only with PyRun) to test something (like virtualenv). If someone need it for miniconda, she/he can implement it.

schlomo commented:
Sure thing. My concern is only to have a single variable for the Python version instead of two. If only the Major part of the Python version applies to miniconda builds and the full version applies to PyRun builds then that is fine with me.

I still would wonder how to get something like "latest 3" or "latest 3.4" with PyRun. IMHO this is a common use case.

sdomme commented:
@schlomo For example look at the default version we use for conda_dist_version. If you search for this in https://repo.continuum.io/miniconda/ there is a Miniconda-3.9.1-xxx and Miniconda3-3.9.1-xxx but nothing with Miniconda2. 
Maybe somebody with more miniconda experience can answer/implement this :-) Somebody can recycle the python_version argument to construct the right thing for miniconda.

sdomme commented:
@schlomo To find latest Python version for PyRun there is some grep-sed-sort-head-magic on Python URL in install-pyrun script from eGenix.

esc commented:
I'm going to merge this for now. One idea I had was to be able to specify the distribution in the yaml file too.
